### PR TITLE
Modify rule S3001: Adapt to LaYC

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**

### CODE REVIEW

File: rules/S3001/javascript/rule.adoc [hunk 1]
```diff
@@ -1,24 +1,29 @@
 == Why is this an issue?
 
-The semantics of the ``++delete++`` operator are a bit tricky, and it can only be reliably used to remove properties from objects. Pass anything else to it, and you may or may not get the desired result.
+The `delete` operator is used to remove a property from an object. It only affects https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn[own] properties. There are two valid forms:
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3001/javascript/rule.adoc:3]:
```suggestion
The `delete` operator is used to remove a property from an object. It only affects its https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn[own] properties. There are two valid ways to remove a property:
```

File: rules/S3001/javascript/rule.adoc [hunk 2]
```diff
@@ -1,24 +1,29 @@
...
+`delete` will throw a `TypeError` in strict mode if the property is a non-configurable property.
+
+`delete identifier` may work if `identifier` is a *configurable* property of the global object. For `identifier` to be *configurable*, it should have been declared directly as a `globalThis` property (`globalThis.identifier = 1`). This form is not common practice and should be avoided. Use `delete globalThis.identifier` instead if needed.
+
+Aside from that case, deleting variables, including function parameters, never works:
+
+* Variables declared with `var` are *non-configurable* properties in the global object, except in CommonJS and ECMAScript modules, where top-level variable declarations are scoped to the module. 
+* Variables declared with `let` or `const` are not attached to any object.
 
 [source,javascript]
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3001/javascript/rule.adoc:17]: Just asking, do we need diff-tag ?

File: rules/S3001/javascript/rule.adoc [hunk 3]
```diff
@@ -28,11 +33,20 @@ var obj = {
   ...
   }
 };
-delete obj.x;
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3001/javascript/rule.adoc:31]: Can you add a space between `:` and `1` above, meaning:

```javascript
...
x: 1,
...
```

Just trying to be consistent here.

File: rules/S3001/javascript/rule.adoc [hunk 4]
```diff
@@ -1,24 +1,29 @@
...
+`delete identifier` may work if `identifier` is a *configurable* property of the global object. For `identifier` to be *configurable*, it should have been declared directly as a `globalThis` property (`globalThis.identifier = 1`). This form is not common practice and should be avoided. Use `delete globalThis.identifier` instead if needed.
+
+Aside from that case, deleting variables, including function parameters, never works:
+
+* Variables declared with `var` are *non-configurable* properties in the global object, except in CommonJS and ECMAScript modules, where top-level variable declarations are scoped to the module. 
+* Variables declared with `let` or `const` are not attached to any object.
 
 [source,javascript]
 ----
 var x  = 1;
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3001/javascript/rule.adoc:19]:
```suggestion
var x = 1;
```

File: rules/S3001/javascript/rule.adoc [hunk 5]
```diff
@@ -1,24 +1,29 @@
 == Why is this an issue?
 
-The semantics of the ``++delete++`` operator are a bit tricky, and it can only be reliably used to remove properties from objects. Pass anything else to it, and you may or may not get the desired result.
+The `delete` operator is used to remove a property from an object. It only affects https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn[own] properties. There are two valid forms:
 
+* `delete object.property`
+* `delete object[property]`
```
Comment (yassin-kammoun-sonarsource) [inline rules/S3001/javascript/rule.adoc:6]:
```suggestion
* Using the dot notation: `delete object.property`
* Using the bracket notation`delete object[property]`
```

**Review by yassin-kammoun-sonarsource — APPROVED** [review]

File: rules/S3001/javascript/rule.adoc (hunk 2 above)
Comment (victor-diez-sonarsource) [reply rules/S3001/javascript/rule.adoc:17]: i didn't add them as the code seems unrelated and all of it will be colored

Comment (yassin-kammoun-sonarsource) [reply rules/S3001/javascript/rule.adoc:17]: 👍

**Pull request merged**